<a href="https://colab.research.google.com/github/jfpIE16/SL_II_Tareas/blob/master/Tarea_III_SLII.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Statistical Learning II
## Tarea 3: Maldición de dimensionalidad

---

José Fernando Pérez Pérez  
josefernando.perez@galileo.edu


In [6]:
from sklearn.preprocessing import PolynomialFeatures
import numpy as np
from tabulate import tabulate

Se propone la utilización de tres conjuntos de datos los cuales representan el caso de 1, 2 y 3 variables de la siguiente forma:  


*   $a$ $\rightarrow$ 1 variable
*   $a,b$ $\rightarrow$ 2 variables
*   $a,b,c$ $\rightarrow$ 3 variables




In [18]:
X_1d = np.random.randn(5,1)
X_2d = np.random.randn(5,2)
X_3d = np.random.randn(5,3)

Utilizando un polinomio de grado 3 se esperarían las siguientes características (incluyendo el sesgo):  

In [27]:
poly_d3 = PolynomialFeatures(degree=3)

**1 Variable**  
$1,a,a²,a³$  
En total 3 características, incluyendo el sesgo

In [32]:
X_t1 = poly_d3.fit_transform(X_1d)
print(tabulate(X_t1))
print(X_t1.shape)

-  ---------  --------  ----------
1  -0.749174  0.561262  -0.420482
1   0.364615  0.132944   0.0484733
1  -0.964034  0.929362  -0.895937
1  -1.25326   1.57066   -1.96844
1  -0.348301  0.121314  -0.0422537
-  ---------  --------  ----------
(5, 4)


**2 Variables**  
$1,a³,b³,a²b,ab²,a²,b²,ab,a,b$  
En total 10 características incluyendo el sesgo

In [31]:
X_t2 = poly_d3.fit_transform(X_2d)
print(tabulate(X_t2))
print(X_t2.shape)

-  ----------  ---------  ----------  ----------  ---------  -----------  -----------  ---------  ----------
1  -0.0903611   0.61555   0.00816513  -0.0556218  0.378901   -0.00073781   0.00502604  -0.034238   0.233233
1  -1.29415     0.502068  1.67481     -0.649749   0.252072   -2.16745      0.840869    -0.326218   0.126557
1   0.712157    0.269028  0.507168     0.19159    0.0723759   0.361183     0.136442     0.051543   0.0194711
1   0.216875   -1.5687    0.0470348   -0.340211   2.46081     0.0102007   -0.0737834    0.533688  -3.86026
1   0.328835   -0.521553  0.108133    -0.171505   0.272018    0.0355579   -0.0563969    0.089449  -0.141872
-  ----------  ---------  ----------  ----------  ---------  -----------  -----------  ---------  ----------
(5, 10)


**3 Variables**  
$1,a³,b³,c³,abc,a²b,a²c,...,a,b,c$  
Y no seguimos porque en total son 20 características generadas

In [30]:
X_t3 = poly_d3.fit_transform(X_3d)
print(tabulate(X_t3))
print(X_t3.shape)

-  ---------  ---------  ---------  ---------  ---------  ---------  --------  --------  ---------  ------------  -----------  ---------  ----------  ---------  ----------  ---------  ---------  ----------  -----------
1   0.844329  -0.680211  -0.342085  0.712891   -0.574322  -0.288832  0.462688  0.23269   0.117022     0.601915     -0.484917   -0.243869    0.390661   0.196467   0.0988049  -0.314725  -0.158278  -0.0795996  -0.0400314
1  -0.145655   1.37412    0.59735   0.0212154  -0.200148  -0.087007  1.88821   0.820832  0.356827    -0.00309012    0.0291525   0.012673   -0.275028  -0.119558  -0.0519736   2.59464    1.12792    0.490324    0.21315
1  -2.69776    0.828904   1.01196   7.2779     -2.23618   -2.73002   0.687081  0.838818  1.02406    -19.634         6.03268     7.36495    -1.85358   -2.26293   -2.76268     0.569524   0.695299   0.84885     1.03631
1  -2.73186   -2.08873   -0.919185  7.46308     5.70611    2.51109   4.36277   1.91992   0.8449     -20.3881      -15.5883     -6.8

**¡Por curiosidad!**  
¿Cuál sería la cantidad de features generadas por un polinomio de grado 3 para un dataset con 200 características de entrada? Los modelos actuales tienen cientos de miles de características, incluso millones como **GTP-3**

In [34]:
X_200d = np.random.randn(5,200)
X_t200 = poly_d3.fit_transform(X_200d)
#print(tabulate(X_t3)) casi, jeje!
print(X_t200.shape)

(5, 1373701)


¡1,373,701! características

**Conclusiones**  
Como se observa, al aumentar tanto las características de entrada como el grado del polinomio observamos un crecimiento exponencial en nuestro conjunto de datos. Aquí radica la importancia de realizar feature engineering como parte necesaria en las aplicaciones de Data Science, sin embargo, optimizar las técnicas de **feature-engineering** siempre van de la mano de la descripción del problema y la familiaridad que el científico de datos tenga con el mismo; podría tomar semanas o meses seleccionar las características y transformaciones correctas que ayuden al algoritmo a converger a una solución.  
Y en esta parte me resulta muy interesante saber que como parte del entrenamiento de algoritmos de **deep learning** se aprenden las características y aplican transformaciones al conjunto de datos que ayuden a resolver el problema de manera automática, claro dependiendo de la cantidad de datos y complejidad de los mismos se puede discutir el costo computacional que esto lleva.